core/lib/zksync_core/src/state_keeper/io/mempool.rs

```diff
@@ -128,15 +161,9 @@ impl StateKeeperIO for MempoolIO {
         max_wait: Duration,
     ) -> Option<(SystemEnv, L1BatchEnv)> {
         let deadline = Instant::now() + max_wait;
-
+        // FIXME: why do we wait for hash immediately and not below? (changed in #809)
         let prev_l1_batch_hash = self.load_previous_l1_batch_hash().await;
```

Comment:
Highlighting this. Not sure whether this influences performance in practice, but theoretically it could be negative. Maybe it makes sense to rename `load_previous_l1_batch_hash` to `wait_for_previous_l1_batch_hash` to be more explicit with what the method does? If this line stays, the "We only need to get the root hash when we're certain that we have a new transaction" comment below should be removed as obsolete. cc @StanislavBreadless

Reply:
The comment seems to not match the logic indeed - probably a bug/leftover. 

> but theoretically it could be negative
@slowli , under which circumstances? From the first glance, we need to wait both for a transaction and for the hash. It doesn't matter in what order to check unless there are heavy operations in betsween

Reply:
Yes, you are right; the order doesn't seem to matter for performance. There are some differences in method semantics though. In the old implementation, once a fitting transaction appears in the mempool, we'll wait for the previous L1 batch hash as long as it takes, and the method is guaranteed to return `Some(_)`. In the new implementation, if we wait for the previous L1 batch hash for more than `max_wait`, the method won't even check the mempool and will immediately return `None`. Now, since the method is only called from [a loop in `ZkSyncStateKeeper`](https://github.com/matter-labs/zksync-era/blob/a8699a20a5e6ff1c6ea99412019f9bfecc3b92f5/core/lib/zksync_core/src/state_keeper/keeper.rs#L258), it doesn't seem to make much difference; the method will presumably get the prev L1 batch hash immediately on the next iteration and will still return `Some(_)` if there are fitting transactions in the mempool. One side-effect of this is that the `wait_for_prev_hash_time` no longer works as intended; it reports bogus near-zero values on these repeated loop iterations.

Reply:
I see. So what would you propose? I'm fine to keep as is (but please remove comment) or change - whatever feels right for you